File: invariance_breaking.ts
class Order {
    protected paid = false;

    ship() {
        if (!this.paid) throw new Error("Not paid");
        console.log("Shipped");
    }
}

class TestOrder extends Order {
    ship() {
        console.log("Shipped"); // rule skipped
    }
}

const o: Order = new TestOrder();
o.ship(); // wrong behavior

File: invariance_breaking.cpp
#include <iostream>
using namespace std;

class Order {
protected:
    bool paid = false;

public:
    virtual void ship() {
        if (!paid) throw runtime_error("Not paid");
        cout << "Shipped\n";
    }
};

class TestOrder : public Order {
public:
    void ship() override {
        cout << "Shipped\n"; // bypasses rule
    }
};

int main() {
    Order* o = new TestOrder();
    o->ship(); // unpaid order shipped
}
